src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexingService.cs

```diff
@@ -0,0 +1,296 @@
+    /// This class provides services to update all the Elasticsearch indices.
+    /// </summary>
+    public class ElasticIndexingService
+    {
+        private const int BatchSize = 100;
+        private readonly IShellHost _shellHost;
+        private readonly ShellSettings _shellSettings;
+        private readonly ElasticIndexSettingsService _elasticIndexSettingsService;
+        private readonly ElasticIndexManager _indexManager;
+        private readonly IIndexingTaskManager _indexingTaskManager;
+        private readonly ISiteService _siteService;
+        private readonly ILogger _logger;
+
+        public ElasticIndexingService(
+            IShellHost shellHost,
+            ShellSettings shellSettings,
+            ElasticIndexSettingsService elasticIndexSettingsService,
+            ElasticIndexManager indexManager,
+            IIndexingTaskManager indexingTaskManager,
+            ISiteService siteService,
+            ILogger<ElasticIndexingService> logger)
+        {
+            _shellHost = shellHost;
+            _shellSettings = shellSettings;
+            _elasticIndexSettingsService = elasticIndexSettingsService;
+            _indexManager = indexManager;
+            _indexingTaskManager = indexingTaskManager;
+            _siteService = siteService;
+            _logger = logger;
+        }
+
+        public async Task ProcessContentItemsAsync(string indexName = default)
+        {
+            var allIndices = new Dictionary<string, long>();
+            var lastTaskId = Int32.MaxValue;
+            IEnumerable<ElasticIndexSettings> indexSettingsList = null;
+
+            if (String.IsNullOrEmpty(indexName))
+            {
+                indexSettingsList = await _elasticIndexSettingsService.GetSettingsAsync();
+
+                if (!indexSettingsList.Any())
+                {
+                    return;
+                }
+
+                // Find the lowest task id to process
+                foreach (var indexSetting in indexSettingsList)
+                {
+                    var taskId = await _indexManager.GetLastTaskId(indexSetting.IndexName);
+                    lastTaskId = Math.Min(lastTaskId, taskId);
+                    allIndices.Add(indexSetting.IndexName, taskId);
+                }
+            }
+            else
+            {
+                var settings = await _elasticIndexSettingsService.GetSettingsAsync(indexName);
+
+                if (settings == null)
+                {
+                    return;
+                }
+
+                indexSettingsList = new ElasticIndexSettings[1] { settings }.AsEnumerable();
+
+                var taskId = await _indexManager.GetLastTaskId(indexName);
+                lastTaskId = Math.Min(lastTaskId, taskId);
+                allIndices.Add(indexName, taskId);
+            }
+
+            if (allIndices.Count == 0)
+            {
+                return;
+            }
+
+            var batch = Array.Empty<IndexingTask>();
+
+            do
```

Comment:
Shouldn't there be a distributed lock around this? Or what will happen if two server nodes index at the same time?

Reply:
Everything is async. It is the background task that would require a lock. Else, if 2 servers initiate manually an index rebuild at the same time then we need to test it.

Reply:
That would be a code optimization. I don't think it will fail because of concurrent indexing. Though, it would optimize the tasks by not getting duplicated in some cases.

Reply:
Waiting for JT, then.

Reply:
I don't think we will need a locking mechanism for the first iteration. I will need to read about the subject and I found interesting examples and posts about it. Generally, you don't need a locking mechanism for Elasticsearch to work in a distributed environment. It could be a code optimization to reduce the load of PUT requests on the server but the background task should generally index only a few items at a time. The locking mechanism becomes interesting if we have a lot of website instances that could potentially send the same task concurrently, but you need to have a lot of INSERTS in your website to need that.

Reply:
Yeah, my questions is more about correctness than performance. I.e. will we see any issues if let's say, 3 instances of the app, on 3 web servers each try to index the same item? You say it won't, OK then.

Reply:
It won't. Elasticsearch is built for that.

Reply:
From what I've read so far the only limitation with Elasticsearch is the amount of data that we push to the Bulk update method which is recommended to not exceed 15 mb at a time. This is why in the ElasticIndexingService we still batch the updates with only 100 content items.

Reply:
Sounds good, then.

Reply:
@piedone @Skrypt 

My "feeling" is that Elastic as a distributed service should use some kind of atomic "transactions", so looks good to not use a lock inside `ProcessContentItems()` itself, but for now make sense to define Lock values for the background task.

A relatively low value for the TimeOut (e.g. 1s), so that an instance don't wait too much before letting the other instance doing the job, and  good enough to get a lock if not yet acquired.

Normally Expiration > max expected time, here the concern is that if it is sometime > 1 min the granularity of bgTasks. But because the indexing is async it would not block the thread pool, only the background service loop would be interrupted, so if it is quite rare that would be okay.

Here it is still better to have Expiration > max expected time (e.g. 5 min). This because anyway a lower value would not interrupt a long indexing, only the lock would expire sooner, and then allowing another process to start another indexing before the previous one was completed.

**So just to say that for me as it is currently done looks for now to be the best compromise.**

Note: When executed from the bgTask one idea would be to limit the processTime, or keep the batchSize of 100 but limit the number of batches that can be done in one bgTask execution.

Note: In fact Locking values depend on the context, for example if we just want to limit the level of concurrency we may want to have a lower Expiration and/or higher TimeOut.

Reply:
There's no locking now, so good then. BTW by "but for now make sense to define Lock values for the background task" do you mean to lock on, i.e. prevent concurrent execution of, the whole background task? I.e. allow `OrchardCore.Search.Elasticsearch.Core.Services.IndexingBackgroundTask` to only run once even if the OC app runs on multiple server nodes?

Reply:
> prevent concurrent execution of, the whole background task?

Yes.

> to only run once even if the OC app runs on multiple server nodes?

Not exactly, only not concurrently (atomicaly), BgTask execution times are not synchronized, may happen at any time in a given minute, so it doesn't prevent that after an indexing another process begin another one just after. But here with the LastId it will see all docs that are already indexed.

Reply:
I see, thanks. Related: https://github.com/OrchardCMS/OrchardCore/issues/11463